{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "gpuType": "T4",
      "authorship_tag": "ABX9TyM7rOvEYD/YUTbilJzlAAcO",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/bennyy117/Salmonella_AMR_Project/blob/main/SalmonellaAMR.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Setup"
      ],
      "metadata": {
        "id": "w0vxOp7S7fjD"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "HS-CJT3te-z2",
        "outputId": "af4b64a7-f536-4eb1-c119-7f478a8b2ea9"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive\n"
          ]
        }
      ],
      "source": [
        "# Mount Google Drive\n",
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.ensemble import RandomForestClassifier\n",
        "from sklearn.multioutput import MultiOutputClassifier\n",
        "from sklearn.metrics import classification_report, accuracy_score\n",
        "import torch\n",
        "import torch.nn as nn\n",
        "import torch.optim as optim\n",
        "from torch.utils.data import Dataset, DataLoader\n",
        "import os\n",
        "\n",
        "data_path = '/content/drive/MyDrive/Salmonella_AMR_Project/'\n",
        "os.makedirs(data_path, exist_ok=True)\n",
        "print(\"Setup complete!\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "rtLU3_7ExIbW",
        "outputId": "2dc75d5a-399e-4f4e-a9e3-8c8d442bcfc1"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Setup complete!\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Load data"
      ],
      "metadata": {
        "id": "Ghd2Im5K7yt8"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "file_name = 'salmonella_amr_data.csv'\n",
        "df = pd.read_csv(data_path + file_name, low_memory=False)\n",
        "\n",
        "print(df.shape)\n",
        "print(df.columns)\n",
        "df.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 533
        },
        "id": "DDmefXWdxN3Q",
        "outputId": "5b89f97f-19ec-4c3e-8194-d0a3f70ced55",
        "collapsed": true
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "(211700, 21)\n",
            "Index(['#Organism group', 'Strain', 'Isolate identifiers', 'Serovar',\n",
            "       'Isolate', 'Create date', 'Location', 'Isolation source',\n",
            "       'Isolation type', 'Food origin', 'SNP cluster', 'Min-same', 'Min-diff',\n",
            "       'BioSample', 'Assembly', 'AMR genotypes', 'Computed types',\n",
            "       'Unnamed: 17', 'Unnamed: 18', 'Unnamed: 19', 'Unnamed: 20'],\n",
            "      dtype='object')\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "       #Organism group      Strain              Isolate identifiers  \\\n",
              "0  Salmonella enterica      SQ0227  \"93-6175B\",\"SQ0227\",\"SRS426868\"   \n",
              "1  Salmonella enterica      SQ0228  \"93-2836A\",\"SQ0228\",\"SRS426867\"   \n",
              "2  Salmonella enterica      SQ0229   \"93-7741\",\"SQ0229\",\"SRS426869\"   \n",
              "3  Salmonella enterica  Gen_001782         \"Gen_001782\",\"SRS426891\"   \n",
              "4  Salmonella enterica  Gen_001783         \"Gen_001783\",\"SRS426892\"   \n",
              "\n",
              "       Serovar         Isolate           Create date             Location  \\\n",
              "0  enteritidis  PDT000000002.3  2014-01-04T17:03:07Z  USA: Western Region   \n",
              "1  enteritidis  PDT000000003.3  2014-01-04T17:03:07Z  USA: Western Region   \n",
              "2  enteritidis  PDT000000004.3  2014-01-04T17:03:07Z  USA: Western Region   \n",
              "3   Heidelberg  PDT000000005.4  2014-01-04T17:03:07Z                  USA   \n",
              "4   Heidelberg  PDT000000006.4  2014-01-04T17:03:07Z                  USA   \n",
              "\n",
              "  Isolation source       Isolation type Food origin  ... Min-same Min-diff  \\\n",
              "0            Ocean  environmental/other         NaN  ...        0      NaN   \n",
              "1            Ocean  environmental/other         NaN  ...        0      NaN   \n",
              "2            Ocean  environmental/other         NaN  ...        0       25   \n",
              "3             food  environmental/other         NaN  ...        2        2   \n",
              "4             food  environmental/other         NaN  ...        2       12   \n",
              "\n",
              "      BioSample         Assembly  \\\n",
              "0  SAMN02147118              NaN   \n",
              "1  SAMN02147119              NaN   \n",
              "2  SAMN02147120              NaN   \n",
              "3  SAMN02147121  GCA_010121905.1   \n",
              "4  SAMN02147122  GCA_010121865.1   \n",
              "\n",
              "                                       AMR genotypes  \\\n",
              "0  ant(2'')-Ia=COMPLETE,aph(3')-Ia=COMPLETE,blaTE...   \n",
              "1                        mdsA=COMPLETE,mdsB=COMPLETE   \n",
              "2                        mdsA=COMPLETE,mdsB=COMPLETE   \n",
              "3         fosA7=COMPLETE,mdsA=COMPLETE,mdsB=COMPLETE   \n",
              "4  blaCMY-2=COMPLETE,fosA7=COMPLETE,mdsA=COMPLETE...   \n",
              "\n",
              "                                 Computed types Unnamed: 17 Unnamed: 18  \\\n",
              "0  antigen_formula=9:g,m:-,serotype=Enteritidis         NaN         NaN   \n",
              "1  antigen_formula=9:g,m:-,serotype=Enteritidis         NaN         NaN   \n",
              "2  antigen_formula=9:g,m:-,serotype=Enteritidis         NaN         NaN   \n",
              "3   antigen_formula=4:r:1,2,serotype=Heidelberg         NaN         NaN   \n",
              "4   antigen_formula=4:r:1,2,serotype=Heidelberg         NaN         NaN   \n",
              "\n",
              "  Unnamed: 19 Unnamed: 20  \n",
              "0         NaN         NaN  \n",
              "1         NaN         NaN  \n",
              "2         NaN         NaN  \n",
              "3         NaN         NaN  \n",
              "4         NaN         NaN  \n",
              "\n",
              "[5 rows x 21 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-ad626e4d-32f4-4846-a71c-db37965d6ddd\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>#Organism group</th>\n",
              "      <th>Strain</th>\n",
              "      <th>Isolate identifiers</th>\n",
              "      <th>Serovar</th>\n",
              "      <th>Isolate</th>\n",
              "      <th>Create date</th>\n",
              "      <th>Location</th>\n",
              "      <th>Isolation source</th>\n",
              "      <th>Isolation type</th>\n",
              "      <th>Food origin</th>\n",
              "      <th>...</th>\n",
              "      <th>Min-same</th>\n",
              "      <th>Min-diff</th>\n",
              "      <th>BioSample</th>\n",
              "      <th>Assembly</th>\n",
              "      <th>AMR genotypes</th>\n",
              "      <th>Computed types</th>\n",
              "      <th>Unnamed: 17</th>\n",
              "      <th>Unnamed: 18</th>\n",
              "      <th>Unnamed: 19</th>\n",
              "      <th>Unnamed: 20</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Salmonella enterica</td>\n",
              "      <td>SQ0227</td>\n",
              "      <td>\"93-6175B\",\"SQ0227\",\"SRS426868\"</td>\n",
              "      <td>enteritidis</td>\n",
              "      <td>PDT000000002.3</td>\n",
              "      <td>2014-01-04T17:03:07Z</td>\n",
              "      <td>USA: Western Region</td>\n",
              "      <td>Ocean</td>\n",
              "      <td>environmental/other</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>SAMN02147118</td>\n",
              "      <td>NaN</td>\n",
              "      <td>ant(2'')-Ia=COMPLETE,aph(3')-Ia=COMPLETE,blaTE...</td>\n",
              "      <td>antigen_formula=9:g,m:-,serotype=Enteritidis</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Salmonella enterica</td>\n",
              "      <td>SQ0228</td>\n",
              "      <td>\"93-2836A\",\"SQ0228\",\"SRS426867\"</td>\n",
              "      <td>enteritidis</td>\n",
              "      <td>PDT000000003.3</td>\n",
              "      <td>2014-01-04T17:03:07Z</td>\n",
              "      <td>USA: Western Region</td>\n",
              "      <td>Ocean</td>\n",
              "      <td>environmental/other</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>SAMN02147119</td>\n",
              "      <td>NaN</td>\n",
              "      <td>mdsA=COMPLETE,mdsB=COMPLETE</td>\n",
              "      <td>antigen_formula=9:g,m:-,serotype=Enteritidis</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Salmonella enterica</td>\n",
              "      <td>SQ0229</td>\n",
              "      <td>\"93-7741\",\"SQ0229\",\"SRS426869\"</td>\n",
              "      <td>enteritidis</td>\n",
              "      <td>PDT000000004.3</td>\n",
              "      <td>2014-01-04T17:03:07Z</td>\n",
              "      <td>USA: Western Region</td>\n",
              "      <td>Ocean</td>\n",
              "      <td>environmental/other</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>25</td>\n",
              "      <td>SAMN02147120</td>\n",
              "      <td>NaN</td>\n",
              "      <td>mdsA=COMPLETE,mdsB=COMPLETE</td>\n",
              "      <td>antigen_formula=9:g,m:-,serotype=Enteritidis</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Salmonella enterica</td>\n",
              "      <td>Gen_001782</td>\n",
              "      <td>\"Gen_001782\",\"SRS426891\"</td>\n",
              "      <td>Heidelberg</td>\n",
              "      <td>PDT000000005.4</td>\n",
              "      <td>2014-01-04T17:03:07Z</td>\n",
              "      <td>USA</td>\n",
              "      <td>food</td>\n",
              "      <td>environmental/other</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>2</td>\n",
              "      <td>2</td>\n",
              "      <td>SAMN02147121</td>\n",
              "      <td>GCA_010121905.1</td>\n",
              "      <td>fosA7=COMPLETE,mdsA=COMPLETE,mdsB=COMPLETE</td>\n",
              "      <td>antigen_formula=4:r:1,2,serotype=Heidelberg</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Salmonella enterica</td>\n",
              "      <td>Gen_001783</td>\n",
              "      <td>\"Gen_001783\",\"SRS426892\"</td>\n",
              "      <td>Heidelberg</td>\n",
              "      <td>PDT000000006.4</td>\n",
              "      <td>2014-01-04T17:03:07Z</td>\n",
              "      <td>USA</td>\n",
              "      <td>food</td>\n",
              "      <td>environmental/other</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>2</td>\n",
              "      <td>12</td>\n",
              "      <td>SAMN02147122</td>\n",
              "      <td>GCA_010121865.1</td>\n",
              "      <td>blaCMY-2=COMPLETE,fosA7=COMPLETE,mdsA=COMPLETE...</td>\n",
              "      <td>antigen_formula=4:r:1,2,serotype=Heidelberg</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows × 21 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-ad626e4d-32f4-4846-a71c-db37965d6ddd')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-ad626e4d-32f4-4846-a71c-db37965d6ddd button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-ad626e4d-32f4-4846-a71c-db37965d6ddd');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-143fcaec-22ab-43c2-abf1-26503b837347\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-143fcaec-22ab-43c2-abf1-26503b837347')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-143fcaec-22ab-43c2-abf1-26503b837347 button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "df"
            }
          },
          "metadata": {},
          "execution_count": 3
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Data processing"
      ],
      "metadata": {
        "id": "A3xTgw5H8Ybq"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Extract unique AMR genes and create binary features\n",
        "def parse_amr_genotypes(row):\n",
        "    if pd.isna(row): return {}\n",
        "    genes = {}\n",
        "    for item in str(row).split(','):\n",
        "        item = item.strip()\n",
        "        if '=' in item:\n",
        "            gene, status = item.split('=')\n",
        "            genes[gene] = 1 if status == 'COMPLETE' else 0\n",
        "    return genes\n",
        "\n",
        "# Apply parsing\n",
        "amr_parsed = df['AMR genotypes'].apply(parse_amr_genotypes)\n",
        "\n",
        "# Get all unique genes\n",
        "all_genes = set()\n",
        "for genes in amr_parsed:\n",
        "    all_genes.update(genes.keys())\n",
        "all_genes = sorted(all_genes)\n",
        "print(f\"Total unique AMR genes: {len(all_genes)}\")\n",
        "\n",
        "# Create binary dataframe\n",
        "amr_df = pd.DataFrame(0, index=df.index, columns=all_genes)\n",
        "for i, genes in enumerate(amr_parsed):\n",
        "    for gene, val in genes.items():\n",
        "        amr_df.at[i, gene] = val\n",
        "\n",
        "# Combine with original (keep only features)\n",
        "features_df = amr_df\n",
        "print(features_df.sum().sort_values(ascending=False).head(20))  # Top common genes"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7UHMC6c0xT9C",
        "outputId": "30170e7f-1adb-4102-954d-83fcea1772bc"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Total unique AMR genes: 429\n",
            "mdsA           204068\n",
            "mdsB           203541\n",
            "aph(6)-Id       25822\n",
            "aph(3'')-Ib     25688\n",
            "tet(A)          24231\n",
            "sul2            21844\n",
            "blaTEM-1        19646\n",
            "tet(B)          16573\n",
            "sul1            15455\n",
            "floR            11705\n",
            "aadA1            9409\n",
            "aadA2            7331\n",
            "aph(3')-Ia       6653\n",
            "blaCMY-2         6492\n",
            "fosA7            5350\n",
            "dfrA14           3773\n",
            "blaCARB-2        3597\n",
            "aac(3)-IVa       3457\n",
            "aph(4)-Ia        3392\n",
            "fosA7.2          3296\n",
            "dtype: int64\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Define key genes for antibiotics (same as before)\n",
        "antibiotic_genes = {\n",
        "    'Ampicillin': ['blaTEM-1', 'blaCMY-2', 'blaCARB-2'],\n",
        "    'Tetracycline': ['tet(A)', 'tet(B)'],\n",
        "    'Streptomycin': ['aph(3\\'\\')-Ib', 'aph(6)-Id', 'aadA1', 'aadA2', 'aph(3\\')-Ia'],\n",
        "    'Sulfonamides': ['sul1', 'sul2'],\n",
        "    'Chloramphenicol': ['floR', 'catA1']\n",
        "}\n",
        "\n",
        "# Create labels (resistant if at least one key gene present)\n",
        "labels_df = pd.DataFrame(0, index=features_df.index, columns=antibiotic_genes.keys())\n",
        "for ab, key_genes_list in antibiotic_genes.items():\n",
        "    available = [g for g in key_genes_list if g in features_df.columns]\n",
        "    if available:\n",
        "        labels_df[ab] = features_df[available].max(axis=1)\n",
        "    resistant_count = labels_df[ab].sum()\n",
        "    prevalence = resistant_count / len(features_df) * 100\n",
        "    print(f\"{ab}: {resistant_count} resistant ({prevalence:.2f}%)\")\n",
        "\n",
        "# Remove key genes from features to avoid direct leakage\n",
        "all_key_genes = set()\n",
        "for genes in antibiotic_genes.values():\n",
        "    all_key_genes.update(genes)\n",
        "key_genes_in_data = [g for g in all_key_genes if g in features_df.columns]\n",
        "\n",
        "features_clean = features_df.drop(columns=key_genes_in_data)\n",
        "print(f\"\\nRemoved {len(key_genes_in_data)} key genes from features.\")\n",
        "print(f\"Remaining features: {features_clean.shape[1]} genes (including common ones like mdsA, mdsB)\")\n",
        "\n",
        "# Final data\n",
        "X = features_clean.values.astype(np.float32)\n",
        "y = labels_df.values.astype(np.float32)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6p0YZeqNy1Os",
        "outputId": "4e56f903-a21e-4c29-9918-884b58d0aafb"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Ampicillin: 28929 resistant (13.67%)\n",
            "Tetracycline: 40003 resistant (18.90%)\n",
            "Streptomycin: 40747 resistant (19.25%)\n",
            "Sulfonamides: 33027 resistant (15.60%)\n",
            "Chloramphenicol: 14621 resistant (6.91%)\n",
            "\n",
            "Removed 14 key genes from features.\n",
            "Remaining features: 415 genes (including common ones like mdsA, mdsB)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Subsample data to speed up training\n",
        "from sklearn.model_selection import train_test_split\n",
        "\n",
        "sample_size = 20000\n",
        "\n",
        "# Stratify on total resistant labels to keep balance\n",
        "X_sub, _, y_sub, _ = train_test_split(X, y, train_size=sample_size, random_state=42, stratify=y.sum(axis=1))\n",
        "\n",
        "print(f\"Subsampled to {X_sub.shape[0]} samples.\")\n",
        "print(\"Resistant prevalence in subsample:\")\n",
        "print(pd.DataFrame(y_sub, columns=antibiotic_genes.keys()).mean() * 100)\n",
        "\n",
        "X = X_sub\n",
        "y = y_sub"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Xewuf8qDy6CA",
        "outputId": "b4557380-bee6-4638-ae93-ad04c7782748"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Subsampled to 20000 samples.\n",
            "Resistant prevalence in subsample:\n",
            "Ampicillin         13.680000\n",
            "Tetracycline       18.925001\n",
            "Streptomycin       19.250000\n",
            "Sulfonamides       15.705000\n",
            "Chloramphenicol     6.755000\n",
            "dtype: float32\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Random Forest with K-Fold\n",
        "\n"
      ],
      "metadata": {
        "id": "63dB8d8w9Q19"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# K-Fold\n",
        "from sklearn.model_selection import StratifiedKFold\n",
        "from sklearn.metrics import accuracy_score\n",
        "import numpy as np\n",
        "\n",
        "# 5-fold CV for Random Forest (multi-label)\n",
        "n_folds = 5\n",
        "skf = StratifiedKFold(n_splits=n_folds, shuffle=True, random_state=42)\n",
        "\n",
        "# Since multi-label, stratify on sum of labels\n",
        "fold_accuracies = {ab: [] for ab in antibiotic_genes.keys()}\n",
        "\n",
        "for fold, (train_idx, test_idx) in enumerate(skf.split(X, y.sum(axis=1))):\n",
        "    X_train, X_test = X[train_idx], X[test_idx]\n",
        "    y_train, y_test = y[train_idx], y[test_idx]\n",
        "\n",
        "    rf = MultiOutputClassifier(RandomForestClassifier(n_estimators=200, random_state=42))\n",
        "    rf.fit(X_train, y_train)\n",
        "\n",
        "    y_pred = rf.predict(X_test)\n",
        "\n",
        "    print(f\"\\nFold {fold+1}:\")\n",
        "    for i, ab in enumerate(antibiotic_genes.keys()):\n",
        "        acc = accuracy_score(y_test[:, i], y_pred[:, i])\n",
        "        fold_accuracies[ab].append(acc)\n",
        "        print(f\"{ab} Accuracy: {acc:.4f}\")\n",
        "\n",
        "# Average accuracies\n",
        "print(\"\\nAverage Accuracies across folds:\")\n",
        "for ab, accs in fold_accuracies.items():\n",
        "    avg_acc = np.mean(accs)\n",
        "    std_acc = np.std(accs)\n",
        "    print(f\"{ab}: {avg_acc:.4f} ± {std_acc:.4f}\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "bGE8-kqlzFYS",
        "outputId": "514e00fc-73ab-4a52-dcec-9d29090accb0"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Fold 1:\n",
            "Ampicillin Accuracy: 0.9147\n",
            "Tetracycline Accuracy: 0.8730\n",
            "Streptomycin Accuracy: 0.8942\n",
            "Sulfonamides Accuracy: 0.9100\n",
            "Chloramphenicol Accuracy: 0.9760\n",
            "\n",
            "Fold 2:\n",
            "Ampicillin Accuracy: 0.9183\n",
            "Tetracycline Accuracy: 0.8690\n",
            "Streptomycin Accuracy: 0.8892\n",
            "Sulfonamides Accuracy: 0.9077\n",
            "Chloramphenicol Accuracy: 0.9690\n",
            "\n",
            "Fold 3:\n",
            "Ampicillin Accuracy: 0.9097\n",
            "Tetracycline Accuracy: 0.8648\n",
            "Streptomycin Accuracy: 0.8938\n",
            "Sulfonamides Accuracy: 0.9117\n",
            "Chloramphenicol Accuracy: 0.9712\n",
            "\n",
            "Fold 4:\n",
            "Ampicillin Accuracy: 0.9105\n",
            "Tetracycline Accuracy: 0.8748\n",
            "Streptomycin Accuracy: 0.8888\n",
            "Sulfonamides Accuracy: 0.9095\n",
            "Chloramphenicol Accuracy: 0.9688\n",
            "\n",
            "Fold 5:\n",
            "Ampicillin Accuracy: 0.9160\n",
            "Tetracycline Accuracy: 0.8672\n",
            "Streptomycin Accuracy: 0.8922\n",
            "Sulfonamides Accuracy: 0.9130\n",
            "Chloramphenicol Accuracy: 0.9695\n",
            "\n",
            "Average Accuracies across folds:\n",
            "Ampicillin: 0.9139 ± 0.0033\n",
            "Tetracycline: 0.8698 ± 0.0037\n",
            "Streptomycin: 0.8916 ± 0.0023\n",
            "Sulfonamides: 0.9104 ± 0.0018\n",
            "Chloramphenicol: 0.9709 ± 0.0027\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# PyTorch with K-Fold"
      ],
      "metadata": {
        "id": "31WrXbaO-1mo"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import torch\n",
        "from torch.utils.data import Dataset, DataLoader\n",
        "import torch.nn as nn\n",
        "import torch.optim as optim\n",
        "from sklearn.model_selection import StratifiedKFold\n",
        "from sklearn.metrics import accuracy_score\n",
        "import numpy as np\n",
        "\n",
        "class AMRDataset(Dataset):\n",
        "    def __init__(self, X, y):\n",
        "        self.X = torch.tensor(X, dtype=torch.float32)\n",
        "        self.y = torch.tensor(y, dtype=torch.float32)\n",
        "\n",
        "    def __len__(self): return len(self.X)\n",
        "    def __getitem__(self, idx): return self.X[idx], self.y[idx]\n",
        "\n",
        "class AMRNet(nn.Module):\n",
        "    def __init__(self, input_size, num_labels):\n",
        "        super().__init__()\n",
        "        self.net = nn.Sequential(\n",
        "            nn.Linear(input_size, 512),\n",
        "            nn.ReLU(),\n",
        "            nn.Dropout(0.3),\n",
        "            nn.Linear(512, 256),\n",
        "            nn.ReLU(),\n",
        "            nn.Dropout(0.3),\n",
        "            nn.Linear(256, num_labels),\n",
        "            nn.Sigmoid()  # For multi-label\n",
        "        )\n",
        "\n",
        "    def forward(self, x):\n",
        "        return self.net(x)\n",
        "\n",
        "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
        "print(\"Using device:\", device)\n",
        "\n",
        "# 5-fold CV for PyTorch\n",
        "n_folds = 5\n",
        "skf = StratifiedKFold(n_splits=n_folds, shuffle=True, random_state=42)\n",
        "\n",
        "fold_accuracies_pt = {ab: [] for ab in antibiotic_genes.keys()}\n",
        "\n",
        "for fold, (train_idx, test_idx) in enumerate(skf.split(X, y.sum(axis=1))):\n",
        "    print(f\"\\nFold {fold+1}:\")\n",
        "\n",
        "    X_train_fold = X[train_idx]\n",
        "    X_test_fold = X[test_idx]\n",
        "    y_train_fold = y[train_idx]\n",
        "    y_test_fold = y[test_idx]\n",
        "\n",
        "    train_dataset = AMRDataset(X_train_fold, y_train_fold)\n",
        "    test_dataset = AMRDataset(X_test_fold, y_test_fold)\n",
        "    train_loader = DataLoader(train_dataset, batch_size=64, shuffle=True)\n",
        "    test_loader = DataLoader(test_dataset, batch_size=64, shuffle=False)\n",
        "\n",
        "    model = AMRNet(X.shape[1], y.shape[1]).to(device)\n",
        "    criterion = nn.BCELoss()\n",
        "    optimizer = optim.Adam(model.parameters(), lr=0.001)\n",
        "\n",
        "    epochs = 20\n",
        "    for epoch in range(epochs):\n",
        "        model.train()\n",
        "        total_loss = 0\n",
        "        for batch_x, batch_y in train_loader:\n",
        "            batch_x, batch_y = batch_x.to(device), batch_y.to(device)\n",
        "            optimizer.zero_grad()\n",
        "            outputs = model(batch_x)\n",
        "            loss = criterion(outputs, batch_y)\n",
        "            loss.backward()\n",
        "            optimizer.step()\n",
        "            total_loss += loss.item()\n",
        "        avg_loss = total_loss / len(train_loader)\n",
        "        if (epoch + 1) % 5 == 0:\n",
        "            print(f\"Epoch {epoch+1}/{epochs}, Loss: {avg_loss:.4f}\")\n",
        "\n",
        "    model.eval()\n",
        "    y_pred_pt = []\n",
        "    y_true_pt = []\n",
        "    with torch.no_grad():\n",
        "        for batch_x, batch_y in test_loader:\n",
        "            batch_x = batch_x.to(device)\n",
        "            outputs = model(batch_x)\n",
        "            y_pred_pt.append((outputs.cpu().numpy() > 0.5).astype(int))\n",
        "            y_true_pt.append(batch_y.numpy())\n",
        "    y_pred_pt = np.vstack(y_pred_pt)\n",
        "    y_true_pt = np.vstack(y_true_pt)\n",
        "\n",
        "    for i, ab in enumerate(antibiotic_genes.keys()):\n",
        "        acc = accuracy_score(y_true_pt[:, i], y_pred_pt[:, i])\n",
        "        fold_accuracies_pt[ab].append(acc)\n",
        "        print(f\"{ab} Accuracy: {acc:.4f}\")\n",
        "\n",
        "# Average accuracies\n",
        "print(\"\\nAverage Accuracies across folds:\")\n",
        "for ab, accs in fold_accuracies_pt.items():\n",
        "    avg_acc = np.mean(accs)\n",
        "    std_acc = np.std(accs)\n",
        "    print(f\"{ab}: {avg_acc:.4f} ± {std_acc:.4f}\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "SN1UdqH0z1dc",
        "outputId": "cd75c530-0b52-4b4f-f148-8db2c3cd8ed4"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Using device: cuda\n",
            "\n",
            "Fold 1:\n",
            "Epoch 5/20, Loss: 0.2778\n",
            "Epoch 10/20, Loss: 0.2713\n",
            "Epoch 15/20, Loss: 0.2682\n",
            "Epoch 20/20, Loss: 0.2668\n",
            "Ampicillin Accuracy: 0.9143\n",
            "Tetracycline Accuracy: 0.8712\n",
            "Streptomycin Accuracy: 0.8958\n",
            "Sulfonamides Accuracy: 0.9075\n",
            "Chloramphenicol Accuracy: 0.9750\n",
            "\n",
            "Fold 2:\n",
            "Epoch 5/20, Loss: 0.2785\n",
            "Epoch 10/20, Loss: 0.2711\n",
            "Epoch 15/20, Loss: 0.2679\n",
            "Epoch 20/20, Loss: 0.2662\n",
            "Ampicillin Accuracy: 0.9187\n",
            "Tetracycline Accuracy: 0.8695\n",
            "Streptomycin Accuracy: 0.8902\n",
            "Sulfonamides Accuracy: 0.9050\n",
            "Chloramphenicol Accuracy: 0.9667\n",
            "\n",
            "Fold 3:\n",
            "Epoch 5/20, Loss: 0.2762\n",
            "Epoch 10/20, Loss: 0.2697\n",
            "Epoch 15/20, Loss: 0.2668\n",
            "Epoch 20/20, Loss: 0.2644\n",
            "Ampicillin Accuracy: 0.9085\n",
            "Tetracycline Accuracy: 0.8640\n",
            "Streptomycin Accuracy: 0.8915\n",
            "Sulfonamides Accuracy: 0.9125\n",
            "Chloramphenicol Accuracy: 0.9712\n",
            "\n",
            "Fold 4:\n",
            "Epoch 5/20, Loss: 0.2763\n",
            "Epoch 10/20, Loss: 0.2702\n",
            "Epoch 15/20, Loss: 0.2663\n",
            "Epoch 20/20, Loss: 0.2645\n",
            "Ampicillin Accuracy: 0.9075\n",
            "Tetracycline Accuracy: 0.8752\n",
            "Streptomycin Accuracy: 0.8885\n",
            "Sulfonamides Accuracy: 0.9075\n",
            "Chloramphenicol Accuracy: 0.9700\n",
            "\n",
            "Fold 5:\n",
            "Epoch 5/20, Loss: 0.2776\n",
            "Epoch 10/20, Loss: 0.2706\n",
            "Epoch 15/20, Loss: 0.2678\n",
            "Epoch 20/20, Loss: 0.2666\n",
            "Ampicillin Accuracy: 0.9165\n",
            "Tetracycline Accuracy: 0.8682\n",
            "Streptomycin Accuracy: 0.8922\n",
            "Sulfonamides Accuracy: 0.9140\n",
            "Chloramphenicol Accuracy: 0.9688\n",
            "\n",
            "Average Accuracies across folds:\n",
            "Ampicillin: 0.9131 ± 0.0044\n",
            "Tetracycline: 0.8697 ± 0.0037\n",
            "Streptomycin: 0.8917 ± 0.0024\n",
            "Sulfonamides: 0.9093 ± 0.0034\n",
            "Chloramphenicol: 0.9704 ± 0.0028\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Model training"
      ],
      "metadata": {
        "id": "e_UJ8Of59-9Q"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "final_dataset = AMRDataset(X, y)\n",
        "final_loader = DataLoader(final_dataset, batch_size=64, shuffle=True)\n",
        "\n",
        "final_model = AMRNet(X.shape[1], y.shape[1]).to(device)\n",
        "optimizer = optim.Adam(final_model.parameters(), lr=0.001)\n",
        "criterion = nn.BCELoss()\n",
        "\n",
        "for epoch in range(30):\n",
        "    final_model.train()\n",
        "    total_loss = 0\n",
        "    for batch_x, batch_y in final_loader:\n",
        "        batch_x, batch_y = batch_x.to(device), batch_y.to(device)\n",
        "        optimizer.zero_grad()\n",
        "        outputs = final_model(batch_x)\n",
        "        loss = criterion(outputs, batch_y)\n",
        "        loss.backward()\n",
        "        optimizer.step()\n",
        "        total_loss += loss.item()\n",
        "    if (epoch + 1) % 10 == 0:\n",
        "        print(f\"Final Epoch {epoch+1}/30, Loss: {total_loss/len(final_loader):.4f}\")\n",
        "\n",
        "torch.save(final_model.state_dict(), data_path + 'final_salmonella_amr_pytorch_model.pth')\n",
        "print(\"Model saved.\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Mr9dVsPn_97V",
        "outputId": "13518a40-a863-4932-995f-62a241ff9a42"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Final Epoch 10/30, Loss: 0.2700\n",
            "Final Epoch 20/30, Loss: 0.2655\n",
            "Final Epoch 30/30, Loss: 0.2641\n",
            "Model saved.\n"
          ]
        }
      ]
    }
  ]
}